File: go/stream_of_characters.go
package leetcode

type StreamChecker struct {
	root   TrieNode
	stream []byte
}

type TrieNode struct {
	node map[byte]TrieNode
}

func Constructor(words []string) StreamChecker {
	streamChecker := StreamChecker{TrieNode{make(map[byte]TrieNode)}, make([]byte, 0, 100)}

	for _, word := range words {
		node := streamChecker.root.node
		for i := len(word) - 1; i >= 0; i-- {
			c := word[i]
			if _, ok := node[c]; !ok {
				node[c] = TrieNode{make(map[byte]TrieNode)}
			}
			node = node[c].node
		}
		if _, ok := node['#']; !ok {
			node['#'] = TrieNode{make(map[byte]TrieNode)}
		}
	}
	return streamChecker
}

func (this *StreamChecker) Query(letter byte) bool {
	this.stream = append(this.stream, letter)
	node := this.root.node

	for i := len(this.stream) - 1; i >= 0; i-- {
		c := this.stream[i]
		if _, ok := node['#']; ok {
			return true
		} else if nextNode, ok := node[c]; ok {
			node = nextNode.node
		} else {
			return false
		}
	}
	if _, ok := node['#']; ok {
		return true
	}
	return false
}

/**
 * Your StreamChecker object will be instantiated and called as such:
 * obj := Constructor(words);
 * param_1 := obj.Query(letter);
 */

File: python/stream-of-characters.py
from typing import List

class StreamChecker:

    def __init__(self, words: List[str]):
        self.rev_trie = self.create_rev_trie(words)
        self.stream = []

    def query(self, letter: str) -> bool:
        self.stream.append(letter)
        return self.find()
    
    def create_rev_trie(self, words):
        root = {}
        for word in words:
            node = root
            for i in range(len(word)-1, -1, -1):
                if word[i] not in node:
                    node[word[i]] = {}
                node = node[word[i]]
            node['#'] = True
        return root
    
    def find(self):
        node = self.rev_trie
        for i in range(len(self.stream)-1, -1, -1):
            if self.stream[i] in node:
                node = node[self.stream[i]]
                if '#' in node: return True
            else:
                return False
        return False